experimental/LieAlgebras/src/root_systems.jl

```diff
@@ -0,0 +1,164 @@
+mutable struct RootSystem <: AbstractVector{fmpq}
+  roots::Vector{Vector{Int}}
+  simple_roots::Vector{Vector{Int}}
+  positive_roots::Vector{Vector{Int}}
+  root_system_type::Tuple{String, Int64}
+  @doc raw"""
+      RootSystem(S::String) -> RootSystem
+
+  Return the root system of type `S` where `S` is a string consisting out of
+  a letter `A`, `B`, `C`, `D`, `E`, `F`, `G` followed by an integer.
+  For the exceptional root system, the integers are fixed, e.g. `G2`, `F4`, `E6`, ...
+  """
+  function RootSystem(S::String)
+    # S is a string detailing the type of the indecomposable root system made out of a letter, e.g. "A", "B", "C",... and an integer for the number of simple roots
+    n = parse(Int64, S[2:end])
+    S1 = GAP.Obj(S[1:1])
+    RS = GAP.Globals.RootSystem(S1, n)
+    Ro_1 = GAP.Globals.PositiveRoots(RS)
+    Ro_2 = GAP.Globals.NegativeRoots(RS)
+    sR = GAP.Globals.SimpleSystem(RS)
+    sR = [[sR[i][j] for j=1:length(sR[i])] for i=1:length(sR)]
+    Ro1 = [[Ro_1[i][j] for j=1:length(Ro_1[i])] for i=1:length(Ro_1)]
+    Ro2 = [[Ro_2[i][j] for j=1:length(Ro_2[i])] for i=1:length(Ro_2)]
+    S = (S[1:1], n)
+    Ro = reduce(vcat, (Ro1, Ro2))
+    return new(Ro,sR,Ro1,S)
+  end
+end
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+ @doc raw"""
+  `size(R::RootSystem, [dim::Int])`
+ Return a tuple containing the dimensions of the vector of roots, i.e. the number of roots in the first entry. 
+ Optionally you can specify a dimension to just get the length of that dimension.
+  """
+size(R::RootSystem, dim::Int) = size(R.roots, dim)
+
+size(R::RootSystem) = size(R.roots)
+
+size(S::String) = size(RootSystem(S))
+
+ @doc raw"""
+  `getindex(R::RootSystem, r::Int)`
+ Return the rth root in the vector of roots in the root system `R`
+  """
+getindex(R::RootSystem, r::Int) = getindex(R.roots, r)
+
+###############################################################################
+#
+#   String I/O
+#
+###############################################################################
+
+function show(io::IO, R::RootSystem)
+  print(io, "Root system of type ")
+  show(io, R.root_system_type)
+end
+
+###############################################################################
+#
+#   further functions
+#
+###############################################################################
+ @doc raw"""
+  `cartan_matrix(S::String) -> Matrix{QQFieldElem}`
+ Return the Cartan matrix of the type of root system specified by `S`
+  """
+function cartan_matrix(S::String)
+  S1 = S[1:1]
+  l = length(S)
+  n = parse(Int64, S[2:l])
```

Comment:
Might be a good idea to put the code which takes a string and "parses" it into a "type" into a separate helper, so it can be shared, and behaves consistently everywhere. This then also makes it easier to improve it later (e.g. we may wish to add input validation in order to get better error messages)